sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSource.scala

```diff
@@ -553,8 +553,12 @@ case class DataSource(
         disallowWritingIntervals(data.schema.map(_.dataType), forbidAnsiIntervals = true)
         SaveIntoDataSourceCommand(data, dataSource, caseInsensitiveOptions, mode)
       case format: FileFormat =>
+        val shouldAllowEmptySchema =
+          sparkSession.sessionState.conf.getConf(SQLConf.ALLOW_EMPTY_SCHEMAS_FOR_WRITES)
         disallowWritingIntervals(data.schema.map(_.dataType), forbidAnsiIntervals = false)
-        DataSource.validateSchema(data.schema)
+        if (!shouldAllowEmptySchema) {
+          DataSource.validateSchema(data.schema)
```

Comment:
Completely skipping `validateSchema` looks wrong to me. If future validations are added there, they will get skipped as well. I think we should pass the conf to `validateSchema`, and then allow `validateSchema` to make a decision about whether or not to check for empty schemas, like:
```
    if (hasEmptySchema(schema) && !conf.getConf(SQLConf.ALLOW_EMPTY_SCHEMAS_FOR_WRITES)) {
      throw QueryCompilationErrors.writeEmptySchemasUnsupportedByDataSourceError()
    }
```